security/tpm_linux_bios_measurements.ksy

```diff
@@ -0,0 +1,434 @@
+meta:
+  id: tpm_linux_bios_measurements
+  title: TPM 2.0 BIOS measurements exposed by Linux kernel
+  license: CC0-1.0
+  endian: le
+doc: |
+  Linux kernel exposes the TPM 2.0 eventlog in a dedicated file,
+  /sys/kernel/security/tpm0/binary_bios_measurements.
+doc-ref: https://trustedcomputinggroup.org/resource/tcg-efi-protocol-specification/
+seq:
+  - id: first_event
+    type: tcg_pcr_event
+    doc: First event of the eventlog
+  - id: events2
+    type: tcg_pcr_event2
+    repeat: eos
+    if: is_agile_log
+    doc: Events of the Crypto Agile Log
+instances:
+  is_agile_log:
+    value: first_event.pcr_index == tcg_pcr_index::pcr0_crtm_and_bios and first_event.event_type == tcg_event_type::no_action and first_event.digest.value == [0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0]
+  specid:
+    pos: 0x20
+    size: first_event.len_event
+    type: tcg_efi_specid_event
+    if: is_agile_log
+    doc: Event Log Header
+  events1:
```

Comment:
What is the purpose of doing it this way instead of just using fields in `seq`?

Reply:
If `events1` was directly in `seq`, it would not contain the first event (it would only be in field `first_event`).
In order to make `events1` contain the first event if is not a Crypto Agile Log header, the only way that I found consisted in using an instance and `pos: 0`. If there is a better way of handling this kind of situation, I will gladly modify this.

Reply:
>it would not contain the first event (it would only be in field first_event).

What's the problem with keeping the first event separate if it is really so special?

Reply:
> What's the problem with keeping the first event separate if it is really so special?

Because it is not special with non-Crypto Agile Log. It is an event that can be used with the other events to recompute the values of the PCR that can be queried from the TPM. For example it is possible to write a program that loops through `events1` and compute _SHA1 hash-extend_ operations for every event for every PCR in order to perform this recomputation. Putting the first event in another field makes such a program more complex.

When the Crypto Agile Log was introduced in TPM 2.0 in order to make it possible to use other hashing algorithms than SHA1, the absence of a proper header in the event log of the TPM became a problem. So the header of the Crypto Agile Log got hacked into the first event of the older format, which is why it is special with the new format.

Reply:
That way making it special.

How about something like this:

<details>

```yaml
meta:
  id: tpm_linux_bios_measurements
  title: TPM 2.0 BIOS measurements exposed by Linux kernel
  license: CC0-1.0
  endian: le
doc: |
  Linux kernel exposes the TPM 2.0 eventlog in a dedicated file,
  /sys/kernel/security/tpm0/binary_bios_measurements.
doc-ref: https://trustedcomputinggroup.org/resource/tcg-efi-protocol-specification/
seq:
  - id: first_event
    type: tcg_pcr_event(0)
    doc: In fact acts like a header
instances:
  is_agile:
    value: first_event.pcr_index == tcg_pcr_index::pcr0_crtm_and_bios and first_event.event_type == tcg_event_type::no_action and first_event.digests.as<sha1>.value == [0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0]
  events:
    pos: 0
    type: tcg_pcr_event(is_agile.to_i)
    repeat: eos
    doc: Events of the Crypto Agile Log except the first one

types:
  tcg_pcr_event:
    doc: SHA1 Event Log Entry
    doc-ref: tdTCG_PCR_EVENT
    params:
      - id: is_agile
        type: u1
    seq:
      - id: pcr_index
        -orig-id: PCRIndex
        type: u4
        enum: tcg_pcr_index
        doc: Platform Configuration Register which is extended
      - id: event_type
        -orig-id: EventType
        type: u4
        enum: tcg_event_type

      - id: digests # we need templates https://github.com/kaitai-io/kaitai_struct/issues/135
        type:
          switch-on: is_agile
          cases:
            1: tcg_pcr_event2_digest
            0: sha1

      - id: len_event
        -orig-id: EventSize
        type: u4
      - id: event
        -orig-id: Event
        size: len_event


  tcg_pcr_event2_digest:
    doc: Crypto Agile Log Entry
    doc-ref: tdTCG_PCR_EVENT2
    seq:
      - id: num_digests
        -orig-id: Count
        type: u4
      - id: digests
        -orig-id: Digests
        type: tpmt_ha
        repeat: expr
        repeat-expr: num_digests
        doc: List of digests extended to

  tpmt_ha:
    doc: Digest with an identified hashing algorithm
    doc-ref: tdTPMT_HA
    seq:
      - id: alg_id
        -orig-id: AlgorithmId
        type: u2
        enum: tpm_alg_id
        doc: Identifier of hashing algorithm
      - id: digest
        -orig-id: Digest
        type:
          switch-on: alg_id
          cases:
            'tpm_alg_id::sha1': sha1
            'tpm_alg_id::sha256': sha256
            'tpm_alg_id::sha384': sha384
            'tpm_alg_id::sha512': sha512

  sha1:
    doc: SHA1 digest
    seq:
      - id: value
        size: 20

  sha256:
    doc: SHA256 digest
    seq:
      - id: value
        size: 32

  sha384:
    doc: SHA384 digest
    seq:
      - id: value
        size: 48

  sha512:
    doc: SHA512 digest
    seq:
      - id: value
        size: 64

  tcg_efi_specid_event:
    doc: TPM 2.0 Event Log Header
    doc-ref: tdTCG_EfiSpecIdEventStruct
    seq:
      - id: magic
        -orig-id: signature
        size: 16
        contents: [Spec ID Event03, 0]
      - id: platform_class
        -orig-id: platformClass
        type: u4
      - id: spec_version_minor
        -orig-id: specVersionMinor
        type: u1
      - id: spec_version_major
        -orig-id: specVersionMajor
        type: u1
      - id: spec_errata
        -orig-id: specErrata
        type: u1
      - id: len_uintn
        -orig-id: uintnSize
        type: u1
        doc: Size of the UINTN fields used in various data structures
      - id: num_algorithms
        -orig-id: numberOfAlgorithms
        type: u4
        doc: Number of hashing algorithms used in this event log (except the first event)
      - id: digest_sizes
        -orig-id: digestSizes
        type: tcg_efi_specid_event_algorithm_size
        repeat: expr
        repeat-expr: num_algorithms
      - id: len_vendor_info
        -orig-id: vendorInfoSize
        type: u1
      - id: vendor_info
        -orig-id: vendorInfo
        size: len_vendor_info
        doc: Provided for use by the BIOS implementer

  tcg_efi_specid_event_algorithm_size:
    doc-ref: tdTCG_EfiSpecIdEventAlgorithmSize
    seq:
      - id: algorithm
        -orig-id: algorithmId
        type: u2
        enum: tpm_alg_id
        doc: TCG defined hashing algorithm ID
      - id: len
        -orig-id: digestSize
        type: u2
        doc: size of the digest for the respective hashing algorithm

enums:
  tcg_pcr_index:
    0:
      id: pcr0_crtm_and_bios
      doc: PCR for CRTM (Core Root of Trust of Measurement), BIOS and Host Platform Extensions
    1:
      id: pcr1_bios_config
      doc: PCR for Host Platform Configuration (including BIOS configuration)
    2:
      id: pcr2_option_rom_code
      doc: PCR for Option ROM Code
    3:
      id: pcr3_option_rom_config
      doc: PCR for Option ROM Configuration and Data
    4:
      id: pcr4_ipl_code
      doc: PCR for Initial Program Load (IPL) Code (eg. the 0x1b8 first bytes of a MBR)
    5:
      id: pcr5_ipl_data
      doc: PCR for Initial Program Load (IPL) Code and Configuration Data
    6:
      id: pcr6_state_transition_and_wake_events
      doc: PCR for State Transition and Wake Events
    7:
      id: pcr7_host_platform_manufacturer_control
    8:
      id: pcr8_operating_system
    9:
      id: pcr9_operating_system
    10:
      id: pcr10_operating_system
    11:
      id: pcr11_operating_system
    12:
      id: pcr12_operating_system
    13:
      id: pcr13_operating_system
    14:
      id: pcr14_operating_system
    15:
      id: pcr15_operating_system
    16:
      id: pcr16_debug
    17:
      id: pcr17_trusted_os_locality_4
    18:
      id: pcr18_trusted_os_locality_3
    19:
      id: pcr19_trusted_os_locality_2
    20:
      id: pcr20_trusted_os_locality_1
    21:
      id: pcr21_trusted_os
    22:
      id: pcr22_trusted_os
    23:
      id: pcr23_application_specific

  tcg_event_type:
    0x00000000:
      id: preboot_cert
      -orig-id: EV_PREBOOT_CERT
    0x00000001:
      id: post_code
      -orig-id: EV_POST_CODE
    0x00000003:
      id: no_action
      -orig-id: EV_NO_ACTION
    0x00000004:
      id: separator
      -orig-id: EV_SEPARATOR
    0x00000005:
      id: action
      -orig-id: EV_ACTION
    0x00000006:
      id: event_tag
      -orig-id: EV_EVENT_TAG
    0x00000007:
      id: s_ctrm_contents
      -orig-id: EV_S_CRTM_CONTENTS
    0x00000008:
      id: s_ctrm_version
      -orig-id: EV_S_CRTM_VERSION
    0x00000009:
      id: cpu_microcode
      -orig-id: EV_CPU_MICROCODE
    0x0000000a:
      id: platform_config_flags
      -orig-id: EV_PLATFORM_CONFIG_FLAGS
    0x0000000b:
      id: table_of_devices
      -orig-id: EV_TABLE_OF_DEVICES
    0x0000000c:
      id: compact_hash
      -orig-id: EV_COMPACT_HASH
    0x0000000d:
      id: ipl
      -orig-id: EV_IPL
    0x0000000e:
      id: ipl_partition_data
      -orig-id: EV_IPL_PARTITION_DATA
    0x0000000f:
      id: nonhost_code
      -orig-id: EV_NONHOST_CODE
    0x00000010:
      id: nonhost_config
      -orig-id: EV_NONHOST_CONFIG
    0x00000011:
      id: nonhost_info
      -orig-id: EV_NONHOST_INFO
    0x00000012:
      id: omit_boot_device_events
      -orig-id: EV_OMIT_BOOT_DEVICE_EVENTS
    0x80000001:
      id: efi_variable_driver_config
      -orig-id: EV_EFI_VARIABLE_DRIVER_CONFIG
    0x80000002:
      id: efi_variable_boot
      -orig-id: EV_EFI_VARIABLE_BOOT
    0x80000003:
      id: efi_boot_services_application
      -orig-id: EV_EFI_BOOT_SERVICES_APPLICATION
    0x80000004:
      id: efi_boot_services_driver
      -orig-id: EV_EFI_BOOT_SERVICES_DRIVER
    0x80000005:
      id: efi_runtime_services_driver
      -orig-id: EV_EFI_RUNTIME_SERVICES_DRIVER
    0x80000006:
      id: efi_gpt_event
      -orig-id: EV_EFI_GPT_EVENT
    0x80000007:
      id: efi_action
      -orig-id: EV_EFI_ACTION
    0x80000008:
      id: efi_platform_firmware_blob
      -orig-id: EV_EFI_PLATFORM_FIRMWARE_BLOB
    0x80000009:
      id: efi_handoff_tables
      -orig-id: EV_EFI_HANDOFF_TABLES
    0x80000010:
      id: efi_hcrtm_event
      -orig-id: EV_EFI_HCRTM_EVENT
    0x800000e0:
      id: efi_variable_authority
      -orig-id: EV_EFI_VARIABLE_AUTHORITY

  tpm_alg_id:
    0x0000:
      id: error
      -orig-id: TPM_ALG_ERROR
    0x0001:
      id: rsa
      -orig-id: TPM_ALG_RSA
    0x0004:
      id: sha1
      -orig-id: TPM_ALG_SHA1
    0x0005:
      id: hmac
      -orig-id: TPM_ALG_HMAC
    0x0006:
      id: aes
      -orig-id: TPM_ALG_AES
    0x0007:
      id: mgf1
      -orig-id: TPM_ALG_MGF1
    0x0008:
      id: keyedhash
      -orig-id: TPM_ALG_KEYEDHASH
    0x000a:
      id: xor
      -orig-id: TPM_ALG_XOR
    0x000b:
      id: sha256
      -orig-id: TPM_ALG_SHA256
    0x000c:
      id: sha384
      -orig-id: TPM_ALG_SHA384
    0x000d:
      id: sha512
      -orig-id: TPM_ALG_SHA512
    0x0010:
      id: "null"
      -orig-id: TPM_ALG_NULL
    0x0012:
      id: sm3_256
      -orig-id: TPM_ALG_SM3_256
    0x0013:
      id: sm4
      -orig-id: TPM_ALG_SM4
    0x0014:
      id: rsassa
      -orig-id: TPM_ALG_RSASSA
    0x0015:
      id: rsaes
      -orig-id: TPM_ALG_RSAES
    0x0016:
      id: rsapss
      -orig-id: TPM_ALG_RSAPSS
    0x0017:
      id: oaep
      -orig-id: TPM_ALG_OAEP
    0x0018:
      id: ecdsa
      -orig-id: TPM_ALG_ECDSA
    0x0019:
      id: ecdh
      -orig-id: TPM_ALG_ECDH
    0x001a:
      id: ecdaa
      -orig-id: TPM_ALG_ECDAA
    0x001b:
      id: sm2
      -orig-id: TPM_ALG_SM2
    0x001c:
      id: ecschnorr
      -orig-id: TPM_ALG_ECSCHNORR
    0x001d:
      id: ecmqv
      -orig-id: TPM_ALG_ECMQV
    0x0020:
      id: kdf1_sp800_56a
      -orig-id: TPM_ALG_KDF1_SP800_56a
    0x0021:
      id: kdf2
      -orig-id: TPM_ALG_KDF2
    0x0022:
      id: kdf1_sp800_108
      -orig-id: TPM_ALG_KDF1_SP800_108
    0x0023:
      id: ecc
      -orig-id: TPM_ALG_ECC
    0x0025:
      id: symcipher
      -orig-id: TPM_ALG_SYMCIPHER
    0x0040:
      id: ctr
      -orig-id: TPM_ALG_CTR
    0x0041:
      id: ofb
      -orig-id: TPM_ALG_OFB
    0x0042:
      id: cbc
      -orig-id: TPM_ALG_CBC
    0x0043:
      id: cfb
      -orig-id: TPM_ALG_CFB
    0x0044:
      id: ecb
      -orig-id: TPM_ALG_ECB
```

</details>

Reply:
```yaml
events:
    pos: 0
    type: tcg_pcr_event(is_agile.to_i)
    repeat: eos
```
This does not work because if the log is a Crypto Agile log, the first "event" (which is a header) follows `tdTCG_PCR_EVENT` structure (with a zeroed SHA1 digest) and all the following events are of type `tdTCG_PCR_EVENT2`.

Reply:
> Since event is usually not parsed (IMHO it should, does its type depend on event_type field?)

Yes. Sometimes it is raw data, sometimes it is `{address, length}` of the measured data (structure `EFI_PLATFORM_FIRMWARE_BLOB`), etc. For more details, this is partially documented in Table 7-1 "EFI Platform Event Types" page 28 of https://trustedcomputinggroup.org/wp-content/uploads/TCG-EFI-Platform-Specification.pdf and I implemented a parser in Python in https://github.com/fishilico/shared/blob/3c295790bb52ee7c2073e85c93649b687e4e9dd1/python/crypto/tpm_decode_bios_measurements.py#L355-L519

I can try implementing parsing the content of the event later and open another Pull Request if it works.

Reply:
>This does not work because if the log is a Crypto Agile log, the first "event" (which is a header) follows tdTCG_PCR_EVENT structure (with a zeroed SHA1 digest) and all the following events are of type tdTCG_PCR_EVENT2.

No problem. Updated the listing.

Reply:
Nice, I did not know about `index`, in `type: tcg_pcr_event(is_agile.to_i & index)`.
However the result seems incompatible with kaitai-struct-visualizer 0.7 (from https://rubygems.org/gems/kaitai-struct-visualizer/versions/0.7, it does not know `params`) and I can't make kaitai-struct-webide working on my host... How can I test your listing?

Reply:
The reason because I want to test is because I am wondering whether the `&` in `type: tcg_pcr_event(is_agile.to_i & index)` should rather be `&&`. For me, `&` means AND bit-by-bit, which would make the types alternate between `tcg_pcr_event(0)` and `tcg_pcr_event(1)` when parsing a Crypto Agile Log. I am nonetheless not sure about this.

Reply:
>`index`

O SHI*, I have forgotten an undescore. `_index` is the right name. Fixed, and `&&` too. Updated the listing.

Reply:
kaitai-struct-visualizer (ksv) is not really "compatible" or "incompatible" with anything per se, as it is using whatever kaitai-struct-compiler (ksc) installed in your system. You can use existing ksv with modern ksc.

Reply:
@GreyCat Indeed. I am using kaitai-struct-compiler version 0.8, which is the latest release on https://github.com/kaitai-io/kaitai_struct/releases. With this, I get the following error:
```sh
$ kaitai-struct-compiler -t ruby tpm_linux_bios_measurements.ksy
/seq/0: parsing expression 'is_agile and  _index).to_i' failed on 1:21, expected "," | CharsWhile(Set( , n)) | "\\\n" | End
```

Which version of the compiler do you consider "modern ksc", if not the latest release?

@KOLANICH is `pos: 0` allowed in an item of a `seq`, like you wrote? I missed this in the documentation, as I thought it was only for `instances`.

Anyway, if you have tested your version of the format with the files I posted, feel free to merge your format instead. I would prefer knowing what "modern ksc" means beforehand, but if you believe I won't be able to understand what your are saying, please close this PR without further comment.

Reply:
>@KOLANICH is pos: 0 allowed in an item of a seq, like you wrote? I missed this in the documentation, as I thought it was only for instances.

I just forgot to move it.

>Anyway, if you have tested your version of the format with the files I posted,

No, I haven't.

`modern ksc` means `nightly`. releases are obsolete and are used mainly for promotion: to create a post on news-sharing websites like LOR, opennet and reddit.

Reply:
> modern ksc means nightly. releases are obsolete and are used mainly for promotion: to create a post on news-sharing websites like LOR, opennet and reddit.

It is sad that the main reason for release is promotion on new-sharing websites. For most projects I work on, releases allow tracking new features (and fixed bugs) across versions, give a signal for packagers to update their packages, and mark a state as "known-good nightly". If I understand correctly, this is for example what youtube-dl does (https://github.com/ytdl-org/youtube-dl/releases), even if this project has very frequent releases.

Anyway, the GitHub Action workflow of `kaitai-io/kaitai_struct_formats` currently uses `kaitai-struct-compiler` version 0.8 from https://dl.bintray.com/kaitai-io/debian/pool/k/kaitai-struct-compiler/: https://github.com/kaitai-io/kaitai_struct_formats/blob/ec44ad14d0128f911bcaa098826edb8a30840cea/.github/workflows/main.yml#L23
If I understood things correctly, you are proposing a format that contained errors and that will break the GitHub Action workflow which is currently configured. What did I miss that would explain that your are not currently to willingly break `kaitai-io/kaitai_struct_formats`?

Reply:
>give a signal for packagers to update their packages

there are official nightlies for Debian

>Anyway, the GitHub Action workflow of kaitai-io/kaitai_struct_formats currently uses kaitai-struct-compiler version 0.8

@GreyCat, should nightly be used?

>If I understood things correctly, you are proposing a format that contained errors

The code I have posted have not been tested anyhow, in fact. I just posted my vision how approximately it should look like. Feel free to take that code and modify it to make it work and look like as you wish. Just keep in mind that parsing the same twice is both performance and memory overhead and that there is such a thing in programming as [DRY principle](https://en.wikipedia.org/wiki/DRY_principle).